From 862aa8ce7dffe0b3e732a433b57e238b068d43cd Mon Sep 17 00:00:00 2001
From: Jason Gerecke <jason.gerecke@wacom.com>
Date: Wed, 31 Jul 2024 15:36:31 -0700
Subject: [PATCH 1/7] Avoid out-of-bounds array read in usbInitToolType and fix
 its fallback

The usbInitToolType function uses `deriveDeviceTypeFromButtonEvent` as
a fallback when it cannot otherwise discover the type of tool being
reported in a packet of kernel events. This fallback was intended
to be called for each event in the packet by iterating over the
`event_ptr` array. The loop is incorrectly implemented, however. Before
we get to this point, the value of `event_ptr` would have already been
incremented to the end of the list by the deviceTypeFromEvent loop.
This means our calls to deriveDeviceTypeFromButtonEvent are accessing
data beyond the valid list of events and possibly even beyond the end
of the array.

To fix this, we can replace the pointer arithmetic in both loops with
array index notation instead.

Fixes: 3eacea1058ef ("Attempt to derive the tool type from a known button/key event")
Signed-off-by: Jason Gerecke <jason.gerecke@wacom.com>
(cherry picked from commit f77696e392e7dff4e5ad6bc989a9fda2fe4fa7db)
---
 src/wcmUSB.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git ./src/wcmUSB.c ../src/wcmUSB.c
index 5b10ee369e8a..6d8770926530 100644
--- ./src/wcmUSB.c
+++ ../src/wcmUSB.c
@@ -1777,9 +1777,9 @@ static int usbInitToolType(WacomDevicePtr priv, int fd,
 {
 	int i, device_type = 0;
 
-	for (i = 0; (i < nevents) && !device_type; ++i, event_ptr++)
+	for (i = 0; (i < nevents) && !device_type; ++i)
 	{
-		device_type = deviceTypeFromEvent(priv, event_ptr->type, event_ptr->code, event_ptr->value);
+		device_type = deviceTypeFromEvent(priv, event_ptr[i].type, event_ptr[i].code, event_ptr[i].value);
 	}
 
 	if (!device_type)
@@ -1789,8 +1789,8 @@ static int usbInitToolType(WacomDevicePtr priv, int fd,
 		device_type = refreshDeviceType(priv, fd);
 
 	if (!device_type) /* expresskey pressed at startup or missing type */
-		for (i = 0; (i < nevents) && !device_type; ++i, event_ptr++)
-			device_type = deriveDeviceTypeFromButtonEvent(priv, event_ptr);
+		for (i = 0; (i < nevents) && !device_type; ++i)
+			device_type = deriveDeviceTypeFromButtonEvent(priv, &event_ptr[i]);
 
 	return device_type;
 }
-- 
2.50.1

